proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+
+This proposal introduces a concept known as package namespaces, allowing a developer to include or exclude package IDs by specifying package namespaces on their package source(s).
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Giving the user more control on what packages are allowed in their software supply chain is an on-going need in the .NET ecosystem. Today, users do not have much control nor capabilities to filter the software dependencies they would consider including in their project. There's a misconception about open source dependencies being more secure which is known as the [Linus' Law](https://en.wikipedia.org/wiki/Linus%27s_Law), which allowing the world an opportunity at your software supply chain is posing a greater risk as more new packages are created everyday. Many requirements by organizations is quite the opposite. Having a controlled list of dependencies that are allowed within the company's policy is regular procedure & considered a best practice.
+
+By providing an allowlist of package namespaces, we believe we can meet the security needs of the ecosystem from cybersquatting & typosquatting attacks like [Dependency Confusion](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610). Additionally, we can extend a security feature that makes NuGet unique of allowing [package creators to reserve a package ID prefix](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation).
+
+This work will also allow future experiences in browsing, installing, and updating packages under an allowed list of namespaces to make developers secure by default when managing their dependencies in their software supply chain.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+When using a combination of public, private, and local sources defined in `NuGet.config` file(s), a user can add a new `<packageNamespaces>` element to opt-in to the feature. They can then create `<packageSource>` children elements to define the feed in which they'd like to add allowed namespaces to. Lastly by adding individual `<namespace>` elements in the `<packageSource>` node, the source will only allow the matching package namespaces from the respective package source.
+
+**Definition:**
+
+Add a new `<packageNamespaces>` element within the `NuGet.config` using the following syntax:
```

Comment:
Are `nuget.exe` and `dotnet` going to support commands for adding and validating these elements?

Reply:
Yes, that's listed under `Future Opportunities`. Same with Visual Studio.